README.md
# jPokeBot

![alt text](https://raw.githubusercontent.com/tassoneroberto/jpokebot/master/logo.png)

## Introduction
:robot: Pokémon GO bot written in Java.

:exclamation: This project is not working with lastest Niantic API versions! :exclamation:

N.B.
This project uses the following APIs: https://github.com/Grover-c13/PokeGOAPI-Java

## Installation
Clone:
```
git clone https://github.com/tassoneroberto/jpokebot.git
```

Add ```PokeGOAPI-library-x.y.z.jar``` file as a project dependency. Lastest versions (0.4.1) is deprecated.

Build maven:
```
mvn package
mvn compile
```

Run ```Bot.java``` class.

## Note
This is an old project decompiled from a JAR file (Yes, I lost original .java files!), so some classes are broken or even missed.
Also, when this bot was created (august 2016), you just needed a Pokemon GO account to login and start botting.
After a while people lost in interest in the Pokemon GO game.
Consequently the developers of the API used by bots and maps stopped to maintain the project updated.
This lead to a difficulty in cracking Niantic game protection.
At the moment doesn't exist working API to use.
Anyway, I decided to upload the project with the hope that one day some API will be available again to use.

## Useful links

https://pogodev.org/

https://github.com/pogodevorg/

https://www.reddit.com/r/pokemongodev/

https://discordapp.com/invite/dKTSHZC/


src/changelog.txt
jPokéBot - CHANGELOG

0.7.8 [13/11/2016]
- Fixed settings reset.
- Removed scan range value "10" from settings.
- Changed folders organization (This version isn't retro compatible).
- Minor changes in check update frame.

0.7.7 [12/11/2016]
- Bot working again.
- Updated libraries.

0.7.6 [27/09/2016]
- Fixed scroll position in "Inventory".
- Fixed displayed text when a Pokémon encounter fails.
- Fixed a Pokémon transferring bug with "Perfect Moves" options selected.
- Fixed a Pokémon transfer bug when trying to transfer similar Pokémon.
- Changed player info displayed from Pokédex to Stardust.

0.7.5 [25/09/2016]
- Implemented auto update function.

0.7.4 [24/09/2016]
- Added "Get level reward" option.

0.7.3 [19/09/2016]
- Added number of candies in Pokébank.

0.7.2 [18/09/2016]
- Added rename by best moves.
- Fixed a bug when catching a Pokémon without Pokéball.

v0.7.1 [17/09/2016]
- Added "Pokémon Rename" option.
- Fixed default recycle kit option.
- Fixed AsyncPokemonGoException at login.
- Fixed old nickname displayed.
- Fixed an item recycle bug.

v0.6.6 [16/09/2016]
- Added Pokémon & Items images.

v0.6.5 [15/09/2016]
- Fixed items and Pokémon names.

v0.6.4 [14/09/2016]
- Fixed some Pokémon transfer bugs.
- Fixed movement speed option.
- Fixed errors in setting files.
- Fixed auto update Inventory and Pokébank.
- Minor GUI fixes (Windows OS).

v0.6.2 [13/09/2016]
- Fixed Pokémon transfer options.
- Fixed login frame freezed.
- Added movement speed in "Pokéstop Loot" options.

v0.6 [12/09/2016]
- Added "Pokémon Transfer" function.
- Added "Pokémon Transfer" options.
- Fixed scan range setting.
- Minor Bot fixes.

v0.5.2 [11/09/2016]
- Added "Pokéstop Loot" options.
- Added "Pokémon Catch" options.
- Fixed auto update item recycle option.
- Minor settings fixes.
- Minor GUI fixes.

v0.5 [10/09/2016]
- Added "Pokémon Catch" function.
- Fixed a bug in items recycling settings.
- Added auto update user info.
- Fixed refresh buttons.
- Minor GUI fixes.

v0.4 [08/09/2016]
- Added "Item Recycle" options.
- Removed unneeded library (reduced JAR size).
- Minor GUI fixes.

v0.3.1 [07/09/2016]
- Improved main frame GUI.
- Changed Bot generated texts.
- Fixed refresh button in Pokébank.

v0.3 [05/09/2016]
- Added "Pokéstop Loot" function.
- Added map scan range.
- Added more coordinates in the login screen.
- Minor GUI fixes.

v0.2 [03/09/2016]
- Added function "Player info".
- Added "Latitude" and "Longitude" input.
- Added save login data option.
- Added default coordinates [Sidney].
- Added sorting option in "Pokébank" by Pokémon's CP, IV, number and name.
- Checked login data to prevent exceptions.
- Bug and GUI fixes.

v0.1 [02/09/2016]
- First release.
- Added login panel.
- Added function "Inventory".
- Added function "Pokébank".

src/bot/CheckUpdate.java
package bot;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLConnection;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JProgressBar;
import javax.swing.SwingUtilities;

public class CheckUpdate extends JFrame {
  private static final long serialVersionUID = 1L;
  private JPanel container;
  private JLabel textNewVersion;
  private JLabel textCurrent;
  protected JLabel textLatest;
  private JButton download;
  private JButton cancel;
  private JProgressBar progress;
  private int currentRelease = 78;
  private int latestRelease;
  private String currentVersion = "0.7.8";
  private String latestVersion;

  public CheckUpdate() {
    setTitle("jPokeBot Update");
    int width = 350;
    int height = 190;
    setSize(width, height);
    setLocationRelativeTo(null);
    setLayout(null);
    setResizable(false);
    setDefaultCloseOperation(3);

    this.container = new JPanel(null);
    add(this.container);
    this.container.setBounds(0, 0, width, height);

    this.textNewVersion = new JLabel("New version detected!");
    this.textLatest = new JLabel("Latest: " + this.latestVersion);
    this.textCurrent = new JLabel("Current: " + this.currentVersion);
    this.download = new JButton("Download");
    this.cancel = new JButton("Cancel");
    this.progress = new JProgressBar();
    this.progress.setStringPainted(true);
    this.progress.setEnabled(false);

    this.container.add(this.textNewVersion);
    this.container.add(this.textLatest);
    this.container.add(this.textCurrent);
    this.container.add(this.progress);
    this.container.add(this.download);
    this.container.add(this.cancel);

    this.textNewVersion.setBounds(10, 10, width - 20, 20);
    this.textLatest.setBounds(10, 40, width - 20, 20);
    this.textCurrent.setBounds(10, 70, width - 20, 20);
    this.progress.setBounds(10, 100, width - 20, 20);
    this.download.setBounds(10, 130, 150, 20);
    this.cancel.setBounds(width - 150 - 10, 130, 150, 20);

    this.download.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent e) {
        CheckUpdate.this.progress.setEnabled(true);
        CheckUpdate.this.download.setEnabled(false);
        CheckUpdate.this.download();
      }
    });

    this.cancel.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent e) {
        FrameLogin fl = new FrameLogin();
        fl.setVisible(true);
        CheckUpdate.this.dispose();
      }
    });
  }

  public static void main(String[] args) throws Exception {
    if (args.length > 0) {
      File toDelete = new File(args[0]);
      toDelete.delete();
    }
    CheckUpdate cu = new CheckUpdate();
    String latest = cu.isUpdated();
    if (latest != null) {
      cu.textLatest.setText("Latest: " + latest);
      cu.setVisible(true);
    } else {
      cu.dispose();
      FrameLogin fl = new FrameLogin();
      fl.setVisible(true);
    }
  }

  private String isUpdated() throws MalformedURLException, IOException, FileNotFoundException {
    URL versionURL = new URL("http://www.jpokebot.altervista.org/latestVersion.html");
    URLConnection yc = versionURL.openConnection();
    BufferedReader in = new BufferedReader(new InputStreamReader(yc.getInputStream()));
    String remote = in.readLine();
    in.close();
    String[] splitLatest = remote.split("\\|");
    this.latestRelease = Integer.parseInt(splitLatest[1]);
    this.latestVersion = splitLatest[0];
    if (this.currentRelease < this.latestRelease)
      return this.latestVersion;
    return null;
  }

  private void download() {
    Runnable updatethread = new Runnable() {
      public void run() {

        try {
          URL url = new URL("http://www.jpokebot.altervista.org/jpokebot_v" + CheckUpdate.this.latestVersion + ".jar");
          HttpURLConnection httpConnection = (HttpURLConnection) url.openConnection();
          long completeFileSize = httpConnection.getContentLength();

          BufferedInputStream in = new BufferedInputStream(httpConnection.getInputStream());
          FileOutputStream fos = new FileOutputStream("jpokebot_v" + CheckUpdate.this.latestVersion + ".jar");
          BufferedOutputStream bout = new BufferedOutputStream(fos, '?');
          byte[] data = new byte[1024];
          long downloadedFileSize = 0L;
          int x = 0;
          while ((x = in.read(data, 0, 1024)) >= 0) {
            downloadedFileSize += x;
            final int currentProgress = (int) (downloadedFileSize / completeFileSize * 100.0D);
            SwingUtilities.invokeLater(new Runnable() {
              public void run() {
                CheckUpdate.this.progress.setValue(currentProgress);
                if (currentProgress == 100) {
                  CheckUpdate.this.downloadCompleted();
                }
              }
            });
            bout.write(data, 0, x);
          }
          bout.close();
          in.close();
        } catch (FileNotFoundException fileNotFoundException) {
        } catch (IOException iOException) {
        }
      }
    };

    (new Thread(updatethread)).start();
  }

  private void downloadCompleted() {
    String toDelete = "";
    try {
      toDelete = CheckUpdate.class.getProtectionDomain().getCodeSource().getLocation().toURI().getPath();
    } catch (URISyntaxException e1) {
      e1.printStackTrace();
    }

    try {
      Runtime.getRuntime().exec("java -jar jpokebot_v" + this.latestVersion + ".jar " + toDelete);
    } catch (IOException e) {

      e.printStackTrace();
    }
    dispose();
  }
}


src/bot/FramePokebank.java
package bot;

import com.pokegoapi.api.PokemonGo;
import com.pokegoapi.exceptions.LoginFailedException;
import com.pokegoapi.exceptions.RemoteServerException;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import javax.swing.ButtonGroup;
import javax.swing.JButton;
import javax.swing.JEditorPane;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JRadioButton;
import javax.swing.JScrollPane;

public class FramePokebank extends JFrame {
  private static final long serialVersionUID = 1L;
  private JPanel container;
  private JLabel textPokebank;
  private JLabel textSort;
  private JButton refresh;
  private JEditorPane display;
  private ButtonGroup sortByGroup;
  private JRadioButton cpButton;
  private JRadioButton ivButton;
  private JRadioButton nButton;
  private JRadioButton azButton;
  private Bot bot;
  private JScrollPane scroll;
  private PokemonGo go;

  public FramePokebank(PokemonGo go) {
    this.go = go;
    this.bot = new Bot(go);
    setTitle("Pokebank");
    int width = 700;
    int height = 600;
    setSize(width, height);
    setLocationRelativeTo(null);
    setLayout(null);
    setResizable(false);
    setDefaultCloseOperation(1);

    this.container = new JPanel(null);
    add(this.container);
    this.container.setBounds(0, 0, width, height);

    this.textSort = new JLabel("Sort by:");
    this.container.add(this.textSort);
    this.textSort.setBounds(10, 10, 50, 20);

    this.textPokebank = new JLabel("Count: " + go.getInventories().getPokebank().getPokemons().size() + "/"
        + go.getPlayerProfile().getPlayerData().getMaxPokemonStorage());
    this.container.add(this.textPokebank);
    this.textPokebank.setBounds(10, 30, 150, 20);

    this.sortByGroup = new ButtonGroup();
    this.cpButton = new JRadioButton("CP", true);
    this.ivButton = new JRadioButton("IV", false);
    this.nButton = new JRadioButton("#", false);
    this.azButton = new JRadioButton("AZ", false);
    this.sortByGroup.add(this.cpButton);
    this.sortByGroup.add(this.ivButton);
    this.sortByGroup.add(this.nButton);
    this.sortByGroup.add(this.azButton);
    this.container.add(this.cpButton);
    this.container.add(this.ivButton);
    this.container.add(this.nButton);
    this.container.add(this.azButton);
    this.cpButton.setBounds(60, 10, 50, 20);
    this.ivButton.setBounds(115, 10, 50, 20);
    this.nButton.setBounds(170, 10, 50, 20);
    this.azButton.setBounds(215, 10, 50, 20);

    this.display = new JEditorPane();
    this.display.setEditable(false);
    this.display.setContentType("text/html");
    this.scroll = new JScrollPane(this.display);
    this.scroll.setVerticalScrollBarPolicy(22);
    this.container.add(this.scroll);
    this.scroll.setBounds(10, 60, width - 20, height - 90);

    this.refresh = new JButton("Refresh");
    this.container.add(this.refresh);
    this.refresh.setBounds(width - 110, 10, 100, 20);
    this.refresh.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent e) {
        FramePokebank.this.refresh();
      }
    });

    refresh();
  }

  public void refresh() {
    String sortBy;
    try {
      this.go.getInventories().updateInventories();
    } catch (LoginFailedException e1) {
      e1.printStackTrace();
    } catch (RemoteServerException e1) {
      e1.printStackTrace();
    }

    if (this.cpButton.isSelected()) {
      sortBy = "cp";
    } else if (this.ivButton.isSelected()) {
      sortBy = "iv";
    } else if (this.azButton.isSelected()) {
      sortBy = "az";
    } else if (this.nButton.isSelected()) {
      sortBy = "number";
    } else {
      sortBy = "default";
    }
    this.display.setText(this.bot.printPokebank(sortBy));
    this.textPokebank.setText("Count: " + this.go.getInventories().getPokebank().getPokemons().size() + "/"
        + this.go.getPlayerProfile().getPlayerData().getMaxPokemonStorage());
    this.display.setCaretPosition(0);
  }
}


src/bot/FrameLogin.java
package bot;

import com.pokegoapi.api.PokemonGo;
import com.pokegoapi.exceptions.AsyncPokemonGoException;
import com.pokegoapi.exceptions.LoginFailedException;
import com.pokegoapi.exceptions.RemoteServerException;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JComboBox;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JPasswordField;
import javax.swing.JTextField;

public class FrameLogin extends JFrame {
  private static final long serialVersionUID = 1L;
  private static HashMap<String, double[]> defaultCoordinates = new HashMap<String, double[]>();
  private JPanel container;
  private JLabel textEmail;
  private JLabel textPassword;
  private JLabel textLatitude;
  private JLabel textLongitude;
  private JLabel loginMessage;
  private JTextField email;
  private JPasswordField password;
  private JTextField latitude;
  private JTextField longitude;
  private JButton login;
  private JButton setCoords;
  private JComboBox<String> defaultCoords;
  private JCheckBox saveData;
  private boolean isAsyncError = true;
  private boolean isWrongInputError = false;

  public FrameLogin() {
    setTitle("jPokeBot");
    int width = 320;
    int height = 270;
    setSize(width, height);
    setLocationRelativeTo(null);
    setLayout(null);
    setResizable(false);
    setDefaultCloseOperation(3);

    this.container = new JPanel(null);
    add(this.container);
    this.container.setBounds(0, 0, width, height);

    this.textEmail = new JLabel("Email");
    this.email = new JTextField();
    this.container.add(this.email);
    this.container.add(this.textEmail);
    this.textEmail.setBounds(10, 10, 300, 20);
    this.email.setBounds(10, 30, 300, 20);

    this.textPassword = new JLabel("Password");
    this.password = new JPasswordField();
    this.container.add(this.password);
    this.container.add(this.textPassword);
    this.textPassword.setBounds(10, 50, 300, 20);
    this.password.setBounds(10, 70, 300, 20);

    this.textLatitude = new JLabel("Latitude");
    this.latitude = new JTextField("");
    this.container.add(this.latitude);
    this.container.add(this.textLatitude);
    this.textLatitude.setBounds(10, 90, 150, 20);
    this.latitude.setBounds(10, 110, 150, 20);

    this.textLongitude = new JLabel("Longitude");
    this.longitude = new JTextField("");
    this.container.add(this.longitude);
    this.container.add(this.textLongitude);
    this.textLongitude.setBounds(160, 90, 150, 20);
    this.longitude.setBounds(160, 110, 150, 20);

    this.setCoords = new JButton("Set coordinates");
    this.container.add(this.setCoords);
    this.setCoords.setBounds(10, 140, 150, 20);

    this.defaultCoords = new JComboBox<String>();
    loadDefaultCoordinates();
    this.container.add(this.defaultCoords);
    this.defaultCoords.setBounds(160, 135, 150, 30);

    this.login = new JButton("Login");
    this.container.add(this.login);
    this.login.setBounds(10, 180, 200, 20);

    this.saveData = new JCheckBox("Save data");
    this.container.add(this.saveData);
    this.saveData.setSelected(true);
    this.saveData.setBounds(220, 177, 100, 25);

    this.loginMessage = new JLabel("Enter your login data to connect.");
    this.container.add(this.loginMessage);
    this.loginMessage.setBounds(10, 210, 300, 20);

    this.setCoords.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent e) {
        double[] coords = (double[]) defaultCoordinates.get(FrameLogin.this.defaultCoords.getSelectedItem());
        FrameLogin.this.latitude.setText(coords[0] + "");
        FrameLogin.this.longitude.setText(coords[1] + "");
      }
    });

    this.login.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent e) {
        FrameLogin.this.login();
      }
    });

    this.email.addKeyListener(new KeyAdapter() {
      public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();
        if (key == 10) {
          FrameLogin.this.login();
        }
      }
    });

    this.password.addKeyListener(new KeyAdapter() {
      public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();
        if (key == 10) {
          FrameLogin.this.login();
        }
      }
    });

    this.latitude.addKeyListener(new KeyAdapter() {
      public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();
        if (key == 10) {
          FrameLogin.this.login();
        }
      }
    });

    this.longitude.addKeyListener(new KeyAdapter() {
      public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();
        if (key == 10) {
          FrameLogin.this.login();
        }
      }
    });

    File dir = new File("jpokebot/");
    if (!dir.exists()) {
      dir.mkdirs();
    }

    File user = new File("jpokebot/user");
    if (!user.exists()) {
      user.mkdirs();
    }

    File changelogFile = new File("jpokebot/changelog.txt");
    if (!changelogFile.exists()) {
      try {
        changelogFile.createNewFile();
        FileWriter fw = new FileWriter(changelogFile.getAbsoluteFile());
        BufferedWriter bw = new BufferedWriter(fw);
        bw.write(getChangelog());
        bw.close();
      } catch (IOException e1) {
        e1.printStackTrace();
      }
    }

    File loginDataFile = new File("jpokebot/user/loginData.txt");
    if (!loginDataFile.exists()) {
      try {
        loginDataFile.createNewFile();
        FileWriter fw = new FileWriter(loginDataFile.getAbsoluteFile());
        BufferedWriter bw = new BufferedWriter(fw);
        bw.write(getDefaultLogin());
        bw.close();
      } catch (IOException e1) {
        e1.printStackTrace();
      }
    } else {
      loadLoginData();
    }

    File defaultRecycleFile = new File("jpokebot/user/settingsItemRecycle.txt");
    if (!defaultRecycleFile.exists()) {
      try {
        defaultRecycleFile.createNewFile();
        FileWriter fw = new FileWriter(defaultRecycleFile.getAbsoluteFile());
        BufferedWriter bw = new BufferedWriter(fw);
        bw.write(getDefaultSettingsRecycle());
        bw.close();
      } catch (IOException e1) {
        e1.printStackTrace();
      }
    }

    File defaultEggHatchFile = new File("jpokebot/user/settingsEggHatch.txt");
    if (!defaultEggHatchFile.exists()) {
      try {
        defaultEggHatchFile.createNewFile();
        FileWriter fw = new FileWriter(defaultEggHatchFile.getAbsoluteFile());
        BufferedWriter bw = new BufferedWriter(fw);
        bw.write(getDefaultEggHatch());
        bw.close();
      } catch (IOException e1) {
        e1.printStackTrace();
      }
    }

    File defaultGeneralFile = new File("jpokebot/user/settingsGeneral.txt");
    if (!defaultGeneralFile.exists()) {
      try {
        defaultGeneralFile.createNewFile();
        FileWriter fw = new FileWriter(defaultGeneralFile.getAbsoluteFile());
        BufferedWriter bw = new BufferedWriter(fw);
        bw.write(getDefaultGeneralSettings());
        bw.close();
      } catch (IOException e1) {
        e1.printStackTrace();
      }
    }

    File defaultPokemonCatchFile = new File("jpokebot/user/settingsPokemonCatch.txt");
    if (!defaultPokemonCatchFile.exists()) {
      try {
        defaultPokemonCatchFile.createNewFile();
        FileWriter fw = new FileWriter(defaultPokemonCatchFile.getAbsoluteFile());
        BufferedWriter bw = new BufferedWriter(fw);
        bw.write(getDefaultSettingsPokemonCatch());
        bw.close();
      } catch (IOException e1) {
        e1.printStackTrace();
      }
    }

    File defaultTransferFile = new File("jpokebot/user/settingsTransfer.txt");
    if (!defaultTransferFile.exists()) {
      try {
        defaultTransferFile.createNewFile();
        FileWriter fw = new FileWriter(defaultTransferFile.getAbsoluteFile());
        BufferedWriter bw = new BufferedWriter(fw);
        bw.write(getDefaultSettingsTransfer());
        bw.close();
      } catch (IOException e1) {
        e1.printStackTrace();
      }
    }

    File defaultRenameFile = new File("jpokebot/user/settingsRename.txt");
    if (!defaultRenameFile.exists()) {
      try {
        defaultRenameFile.createNewFile();
        FileWriter fw = new FileWriter(defaultRenameFile.getAbsoluteFile());
        BufferedWriter bw = new BufferedWriter(fw);
        bw.write(getDefaultSettingsRename());
        bw.close();
      } catch (IOException e1) {
        e1.printStackTrace();
      }
    }
  }

  private void loadDefaultCoordinates() {
    double[] sidneyCoords = { -33.86302591709663D, 151.2098517268896D };
    defaultCoordinates.put("[AU] Sidney", sidneyCoords);

    double[] hongKongCoords = { 22.285661566028704D, 114.15020227432251D };
    defaultCoordinates.put("[CN] Hong Kong", hongKongCoords);

    double[] tokyoCoords = { 35.684895044528794D, 139.70996975898743D };
    defaultCoordinates.put("[JP] Tokyo", tokyoCoords);

    double[] newYorkCoords = { 40.7711329D, -73.9741874D };
    defaultCoordinates.put("[US] New York", newYorkCoords);

    double[] santaMonicaCoords = { 34.01478835307631D, -118.4959434915362D };
    defaultCoordinates.put("[US] Santa Monica", santaMonicaCoords);

    double[] californiaCoords = { 34.0092419D, -118.49760370000001D };
    defaultCoordinates.put("[US] California", californiaCoords);

    double[] berlinCoords = { 52.537023D, 13.203011D };
    defaultCoordinates.put("[EU] Berlin", berlinCoords);

    double[] milanCoords = { 45.4721D, 9.17722D };
    defaultCoordinates.put("[EU] Milan", milanCoords);

    for (String location : defaultCoordinates.keySet()) {
      this.defaultCoords.addItem(location);
    }
  }

  private String getDefaultLogin() {
    BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("/res/config/defaultLoginData.txt")));
    StringBuilder sb = new StringBuilder();
    try {
      String line;
      while ((line = br.readLine()) != null) {
        sb.append(String.valueOf(line) + System.lineSeparator());
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return sb.toString();
  }

  private String getDefaultSettingsRecycle() {
    BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("/res/config/defaultSettingsItemRecycle.txt")));
    StringBuilder sb = new StringBuilder();
    try {
      String line;
      while ((line = br.readLine()) != null) {
        sb.append(String.valueOf(line) + System.lineSeparator());
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return sb.toString();
  }

  private String getDefaultGeneralSettings() {
    BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("/res/config/defaultSettingsGeneral.txt")));
    StringBuilder sb = new StringBuilder();
    try {
      String line;
      while ((line = br.readLine()) != null) {
        sb.append(String.valueOf(line) + System.lineSeparator());
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return sb.toString();
  }

  private String getDefaultSettingsPokemonCatch() {
    BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("/res/config/defaultSettingsPokemonCatch.txt")));
    StringBuilder sb = new StringBuilder();
    try {
      String line;
      while ((line = br.readLine()) != null) {
        sb.append(String.valueOf(line) + System.lineSeparator());
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return sb.toString();
  }

  private String getDefaultSettingsTransfer() {
    BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("/res/config/defaultSettingsTransfer.txt")));
    StringBuilder sb = new StringBuilder();
    try {
      String line;
      while ((line = br.readLine()) != null) {
        sb.append(String.valueOf(line) + System.lineSeparator());
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return sb.toString();
  }

  private String getDefaultSettingsRename() {
    BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("/res/config/defaultSettingsRename.txt")));
    StringBuilder sb = new StringBuilder();
    try {
      String line;
      while ((line = br.readLine()) != null) {
        sb.append(String.valueOf(line) + System.lineSeparator());
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return sb.toString();
  }

  private String getDefaultEggHatch() {
    BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("/res/config/defaultSettingsEggHatch.txt")));
    StringBuilder sb = new StringBuilder();
    try {
      String line;
      while ((line = br.readLine()) != null) {
        sb.append(String.valueOf(line) + System.lineSeparator());
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return sb.toString();
  }

  private String getChangelog() {
    BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("/changelog.txt")));
    StringBuilder sb = new StringBuilder();
    try {
      String line;
      while ((line = br.readLine()) != null) {
        sb.append(String.valueOf(line) + System.lineSeparator());
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
    return sb.toString();
  }

  protected String checkLoginData() {
    if (this.email.getText().equals("") || this.password.getPassword().toString().equals("")
        || this.latitude.getText().equals("") || this.longitude.getText().equals("")) {
      return "Error: all fields are required.";
    }
    boolean coordinatesOK = true;
    try {
      Double.parseDouble(this.latitude.getText());
      Double.parseDouble(this.longitude.getText());
    } catch (NumberFormatException e) {
      coordinatesOK = false;
    }
    if (!coordinatesOK) {
      return "Error: check your coordinates.";
    }
    return "";
  }

  private void loadLoginData() {
    try {
      BufferedReader br = new BufferedReader(
        new InputStreamReader(getClass().getResourceAsStream("jpokebotuser/loginData.txt")));
      String line;
      while ((line = br.readLine()) != null) {
        String[] info = line.split("=");
        if (info.length == 2 && !info[1].equals("")) {
          if (info[0].equals("email")) {
            this.email.setText(info[1]);
            continue;
          }
          if (info[0].equals("password")) {
            this.password.setText(info[1]);
            continue;
          }
          if (info[0].equals("latitude")) {
            this.latitude.setText(info[1]);
            continue;
          }
          if (info[0].equals("longitude")) {
            this.longitude.setText(info[1]);
            continue;
          }
          if (info[0].equals("saveLoginData") && info[1].equals("true")) {
            this.saveData.setSelected(true);
          }
        }
      }
    } catch (IOException | NullPointerException e) {
      e.printStackTrace();
    }
  }

  protected void login() {
    this.isAsyncError = true;
    this.isWrongInputError = false;
    LoginLoader loader = new LoginLoader();
    disableButtons();
    this.loginMessage.setText("Connecting...");
    loader.start();
  }

  private void disableButtons() {
    this.login.setEnabled(false);
    this.email.setEnabled(false);
    this.password.setEnabled(false);
    this.latitude.setEnabled(false);
    this.longitude.setEnabled(false);
    this.setCoords.setEnabled(false);
    this.defaultCoords.setEnabled(false);
    this.saveData.setEnabled(false);
  }

  private void enableButtons() {
    this.login.setEnabled(true);
    this.email.setEnabled(true);
    this.password.setEnabled(true);
    this.latitude.setEnabled(true);
    this.longitude.setEnabled(true);
    this.setCoords.setEnabled(true);
    this.defaultCoords.setEnabled(true);
    this.saveData.setEnabled(true);
  }

  public class LoginLoader extends Thread {
    public void run() {
      while (FrameLogin.this.isAsyncError && !FrameLogin.this.isWrongInputError) {
        String loginDataError = FrameLogin.this.checkLoginData();
        if (!loginDataError.equals("")) {
          FrameLogin.this.loginMessage.setText(loginDataError);
          FrameLogin.this.isWrongInputError = true;
          continue;
        }
        FrameLogin.this.isWrongInputError = false;
        FrameLogin.this.isAsyncError = false;
        PokemonGo go = null;
        try {
          go = Bot.connect(FrameLogin.this.email.getText(), new String(FrameLogin.this.password.getPassword()));
        } catch (LoginFailedException e1) {
          FrameLogin.this.loginMessage.setText("Wrong input data!");
          FrameLogin.this.enableButtons();
          FrameLogin.this.isWrongInputError = true;
        } catch (RemoteServerException e1) {
          FrameLogin.this.loginMessage.setText("Server unreachable! Retrying...");
          FrameLogin.this.isAsyncError = true;
        } catch (AsyncPokemonGoException e1) {
          FrameLogin.this.loginMessage.setText("Async exception! Retrying...");
          FrameLogin.this.isAsyncError = true;
        }
        if (!FrameLogin.this.isAsyncError && !FrameLogin.this.isWrongInputError) {
          if (FrameLogin.this.saveData.isSelected())
            FrameLogin.this.saveLoginData();
          FrameLogin.this.inzializeBot(go, Double.parseDouble(FrameLogin.this.latitude.getText()),
              Double.parseDouble(FrameLogin.this.longitude.getText()));
        }
      }
    }
  }

  private void saveLoginData() {
    StringBuilder sb = new StringBuilder();
    sb.append("jPokeBot - LOGIN DATA" + System.lineSeparator() + System.lineSeparator());
    sb.append("email=" + this.email.getText() + System.lineSeparator());
    sb.append("password=" + new String(this.password.getPassword()) + System.lineSeparator());
    sb.append("latitude=" + this.latitude.getText() + System.lineSeparator());
    sb.append("longitude=" + this.longitude.getText() + System.lineSeparator());
    sb.append("saveLoginData=true");

    File loginDataFile = new File("jpokebotuser/loginData.txt");

    try {
      FileWriter fw = new FileWriter(loginDataFile.getAbsoluteFile());
      BufferedWriter bw = new BufferedWriter(fw);
      bw.write(sb.toString());
      bw.close();
    } catch (IOException e) {
      e.printStackTrace();
    }
  }

  protected void inzializeBot(PokemonGo go, double latitude, double longitude) {
    setVisible(false);
    FrameBot fb = new FrameBot(go, latitude, longitude);
    fb.setVisible(true);
  }

  public static void main(String[] args) {
    JFrame f = new FrameLogin();
    f.setVisible(true);
  }
}
